### ISSUE TITLE
Hi

### ISSUE THREAD

**Issue by JOJ0** [mgoltzsche/beets-autogenre#1]
How could you hide that plugin from me for more than 3 months?? ;-))) The descriptions sound awesome and exactly what I'm looking for. I'll read some code and play around with it and maybe add some more details to the docs once I find out what it does! It sounds superpowerful! Thanks again and sorry for the spam issue! Only problem is I'm travelling currently with limited computing resources 😁

**Comment by mgoltzsche** [comment]
Great that you find it useful too!
FYI: Unfortunately, for genre detection xtractor/essentia turned out not to be very useful to me: After I tweaked my lastgenre plugin config a bit, I got better results: In case of my library only 8/3k songs didn't have a lastgenre-assigned genre. Unfortunately xtractor/essentia doesn't help in this case since it returns the `dance` genre for these 8 songs :disappointed:. The `dance` genre seems to be where songs of various genres end up when it couldn't detect the genre. Therefore, I am thinking about treating the `dance` genre as not found/empty string. For other songs the public essentia models mostly detected the correct genre but they are less granular than last.fm tags so that it would not be a good replacement for the last.fm integration.
Also, I started working on supporting an `auto` config option just like the lastgenre plugin has.

**Comment by JOJ0** [comment]
Interesting! Thanks for sharing your usecases and what works for you. I have quite some tracks that are not to found via lastgenre. No access to my library currently, but the music I have on my mobile via Navidrome-offline tells me already that there is quite some (I have a fallback genre configured named NoLastGenre):
![image](https://github.com/mgoltzsche/beets-autogenre/assets/2733783/02e854f1-082b-42a6-b577-46cf38a20729)

What I sometimes do is run whatlastgenre over those albums, which fixes at least some of them. The problem with the latter plugin though is that it's only album-based and not track-based as I prefer to tag my genres.

I'd love to tinker with Essentia genre-rosamerica now to see if those NoLastGenre tracks would get anything more zseful than "dance" for my lib! But too bad no lib for the next couple months ;-)

Very interesting topic in general: getting genres in a lib right. Do you mind if we keep this issue open as a general "sharing usecases and findings chitchat" thread?

**Comment by mgoltzsche** [comment]
I forgot to mention that the lastgenre plugin didn't find the genre for ~300 songs in my library (and found the wrong one for a few). However, it turned out that in my library most of those songs were remixes/sets/collections and the genre was part of the song or album title (the ytimport plugin splits mixes/long tracks with chapter metadata and imports them as albums). That's why I made the autogenre plugin derive the genre from the title:
![grafik](https://github.com/mgoltzsche/beets-autogenre/assets/13354530/cf3c10e3-3962-4fc2-a6cd-1ba7f1ba28ef)

Btw the newer songs within my library with an essentia-based genre have a more meaningful genre assigned (not `dance`) but unfortunately not as granular as the title and lastfm based genres within my library:
![grafik](https://github.com/mgoltzsche/beets-autogenre/assets/13354530/60ec26da-c54e-4fe2-84d5-717f43f4f1ce)

> Do you mind if we keep this issue open as a general "sharing usecases and findings chitchat" thread?

Sure, we can do that.

**Comment by mgoltzsche** [comment]
It would be great if you could shed some light on [the odd Beets behaviour](https://github.com/mgoltzsche/beets-autogenre/pull/2#discussion_r1616446765) I discovered while trying to make the autogenre plugin run on import automatically within #2.

Also, I wonder what makes you favour Navidrome-offline over other mobile players? I didn't try using Navidrome yet.

**TL;DR:**
In case you're curious about my actual Beets use-cases, here they are:
* As part of my multi-room audio system at home I have Beets integrated with [Mopidy](https://github.com/mopidy/mopidy) (see [beets-container](https://github.com/mgoltzsche/beets-container) and [mopidy-container](https://github.com/mgoltzsche/mopidy-container)), primarily to play the playlists generated by [my smartplaylist plugin rules](https://github.com/mgoltzsche/beets-container/blob/v0.16.0/config.yaml#L319) which are mostly genre-based. It is all running as apps on my experimental Kubernetes distribution ([kubemate](https://github.com/mgoltzsche/kubemate)) on several Raspberry Pis with Hifiberry HATs.
  * Kubemate start screen (onboarding UX needs to be polished before folks can use it really): ![Screenshot_2024-05-28-02-33-06-188_com android chrome](https://github.com/mgoltzsche/beets-autogenre/assets/13354530/faec7509-06a7-40fb-9a55-ad31f56c991d)
  * Mopidy Iris (playing from Beets via Subsonic/[Beetstream](https://github.com/BinaryBrain/Beetstream)/Mopidy-Subidy but using playlists from the [beets-webm3u plugin](https://github.com/mgoltzsche/beets-webm3u)/Mopidy-WebM3U while [Iris doesn't display Subidy-based playlists yet](https://github.com/jaedb/Iris/pull/947)): ![Screenshot_2024-05-28-01-10-42-985_com android chrome](https://github.com/mgoltzsche/beets-autogenre/assets/13354530/4c9baa0c-44a0-4c57-8b9c-303d518c8e6a)

* To listen to music offline on my mobile phone, I use the VLC player app and regularly rsync my beets library to the phone, including the generated M3U playlists. 
![Screenshot_2024-05-28-01-19-44-757_org videolan vlc](https://github.com/mgoltzsche/beets-autogenre/assets/13354530/c4c1a56a-281a-47b6-a55e-00d20b91b6fd)

Before I used Beets, I used Mopidy with the [Mopidy-YouTube](https://github.com/natumbri/mopidy-youtube) plugin to listen to music - I can still highly recommend it since its auto-play feature finds related, potentially new music which is a feature that is still missing in my Beets setup. When I went to an off-grid vacation last year I wanted to bring some music with me for which I wrote a primitive shell script spontaneously to copy my Mopidy-Youtube cache from the Raspberry Pi, generate a playlist from it and copy it to the phone. Obviously the playlist wasn't great because it was a mix of everything - but better than nothing. When I came back from vacation I came across Beets and started playing with it.

**Comment by JOJ0** [comment]
> I wonder what makes you favour Navidrome-offline over other mobile players? I didn't try using Navidrome yet.
> 

Other mobile players? Not sure if I understand entirely correct? Or did you want to say "other media servers instead of navidrome?"

Or is the question why I dont simply copy over mp3 files to my mobile without any server?

**Comment by mgoltzsche** [comment]
> music I have on my mobile via Navidrome-offline

It sounded like you are using a Navidrome app on your mobile phone to play music but I may have misunderstood that since I don't find a Navidrome mobile app. Though, then I wonder how are you playing the music on your mobile phone offline while you're travelling when you don't just copy the files over? Which kind of app is it you made a screenshot of? Is it a Subsonic client with an offline mode? You don't bring a Raspberry Pi with you while you're travelling and you are not running the Navidrome server on your phone, are you :smiley:?
My copy-all-files-approach may not scale well with my growing music library and I already encountered problems with the MTP file synchronization implementation on Ubuntu.
In case you already tried multiple of these apps and offline music playing workflows and think that Navidrome and this app works best for you, I'd be curious to hear about that. Otherwise don't bother writing a roman about it :wink:.

My [other question](https://github.com/mgoltzsche/beets-autogenre/pull/2#discussion_r1616446765) is actually more important: why is Beets failing to write to custom fields during import but succeeding to writes those fields when the same code is run via a separate command (`beet autogenre`)?

**Comment by JOJ0** [comment]
> > music I have on my mobile via Navidrome-offline
> 
> 
> 
> It sounded like you are using a Navidrome app on your mobile phone to play music but I may have misunderstood that since I don't find a Navidrome mobile app. Though, then I wonder how are you playing the music on your mobile phone offline while you're travelling when you don't just copy the files over? Which kind of app is it you made a screenshot of? Is it a Subsonic client with an offline mode? You don't bring a Raspberry Pi with you while you're travelling and you are not running the Navidrome server on your phone, are you :smiley:?
> 
> My copy-all-files-approach may not scale well with my growing music library and I already encountered problems with the MTP file synchronization implementation on Ubuntu.
> 
> In case you already tried multiple of these apps and offline music playing workflows and think that Navidrome and this app works best for you, I'd be curious to hear about that. Otherwise don't bother writing a roman about it :wink:.


No worries, just wanted to clarify what you are interested in. 🙂 

Yeah that was sloppy wording. I currently use play:sub subsonic client on an iphone. I'm not always happy with that app because of tiny issues and the dev is not responding at all to my suggestions, questions, friendly messages. It's a paid app. Still, it's one of the best subsonic apps on mobile.

I want to switch to Amperfy app. FOSS and dev responsive, just some features missing I'm waiting for to suit my needs.

navidrome is the server because I want to run it directly on my dated 2010 Synology NAS. It's quite minimalist and so far the only server that performs well on that hardware. Dev is responsive and creates wuality code it seems.

copying over files to a filesystem that is accessible from apps is not a thing on iphones. Also 128GB storage is not enough for my library;-)

I simply wanted a flexible solution to make albums and beets dmartplaylists available on my phone, similar to streaming spps like spotify. one button press should make things offline or not. server transcodes my flacs to 320 mp3 to save space on phone. my server is not yet on the internet. maybe I'll find the time to securely implement that one day. it's probably only a docker https relay away ;-)

**Comment by JOJ0** [comment]
> 
> My [other question](https://github.com/mgoltzsche/beets-autogenre/pull/2#discussion_r1616446765) is actually more important: why is Beets failing to write to custom fields during import but succeeding to writes those fields when the same code is run via a separate command (`beet autogenre`)?

I don't have an answer sorry, only some probably useless ideas on how to investigate...

Is genres a db field? Or a flexible attribute?

Is genres a proper nullchar delimited field as implemented in #fixme

Or was it there before that in beets? Without any magic it would be a simple flexible attribute. in theory

The error points to beets trying to access it via an actual sql statement, which is just wrong. somewhere higher in the code it should realize that this is not a fixed db field and only a flex attr.

maybe you look up where that magic happens snd what could be the difference when it's run during import

not much help I'm sorry...

**Comment by mgoltzsche** [comment]
Thanks for letting me know about your setup!
I am an Android user so that I cannot use the Amperfy app unfortunately but the sync mechanism you describe sounds like what would scale with a larger library, indeed, and what I could also use.
I didn't expose my music library via HTTP within the internet (yet). I worked on the Beets HTTP support in order to decouple it from Mopidy so that both apps/docker containers could be installed separately which works best via network in the kubernetes/microservice/container world, even though in practice both may run on the same machine. That said, I am a bit surprised that Beets' Subsonic server plugin (Beetstream) isn't more popular.


I answered your questions related to the attribute error on import within the PR now.
